### ISSUE TITLE
Quarkus random ConcurrentModificationException 

### ISSUE THREAD

**Issue by silviu-negoita** [quarkusio/quarkus#17322]
## Issue 
After upgrading from `1.12.2.Final` to `1.13.4.Final` a lot of **random** `java.util.ConcurrentModificationException`  are thrown when a rest API call is done. For example, it happens between FE(Angular) and BE communication, but also between micro-services(all Quarkus based). This makes local environment(dev mode) almost unusable. 

**P.S. This is reproducible only on local, running the dev mode.**

Is something changed in this direction between mentioned Quarkus version? How can I debug it further? 

Please feel free to ask for more information.

## Example
Here is an example, when FE calls the BE `/myc/api/user/me`
```
021-05-18 10:08:05,388 ERROR [io.und.req.io] (executor-thread-37) Exception handling request e981e92b-df4b-4526-8b03-fd23e4e08faf-5 to /myc/api/user/me: java.util.ConcurrentModificationException
	at java.base/java.util.ArrayList$Itr.checkForComodification(ArrayList.java:1043)
	at java.base/java.util.ArrayList$Itr.next(ArrayList.java:997)
	at io.quarkus.arc.runtime.devconsole.Invocation$Builder.build(Invocation.java:173)
	at io.quarkus.arc.runtime.devconsole.Invocation$Builder.build(Invocation.java:174)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor.proceed(InvocationInterceptor.java:71)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor.monitor(InvocationInterceptor.java:49)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor_Bean.intercept(InvocationInterceptor_Bean.zig:521)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:41)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:41)
	at io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:32)
	at com.connectis.myconnectis.adapters.web.infra.filters.AngularResourcesFilter_Subclass.doFilter(AngularResourcesFilter_Subclass.zig:1090)
	at com.connectis.myconnectis.adapters.web.infra.filters.AngularResourcesFilter_ClientProxy.doFilter(AngularResourcesFilter_ClientProxy.zig:444)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at io.undertow.servlet.handlers.FilterHandler.handleRequest(FilterHandler.java:84)
	at io.undertow.servlet.handlers.security.ServletSecurityRoleHandler.handleRequest(ServletSecurityRoleHandler.java:63)
	at io.undertow.servlet.handlers.ServletChain$1.handleRequest(ServletChain.java:68)
	at io.undertow.servlet.handlers.ServletDispatchingHandler.handleRequest(ServletDispatchingHandler.java:36)
	at io.undertow.servlet.handlers.RedirectDirHandler.handleRequest(RedirectDirHandler.java:67)
	at io.undertow.servlet.handlers.security.SSLInformationAssociationHandler.handleRequest(SSLInformationAssociationHandler.java:133)
	at io.undertow.servlet.handlers.security.ServletAuthenticationCallHandler.handleRequest(ServletAuthenticationCallHandler.java:57)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.security.handlers.AbstractConfidentialityHandler.handleRequest(AbstractConfidentialityHandler.java:46)
	at io.undertow.servlet.handlers.security.ServletConfidentialityConstraintHandler.handleRequest(ServletConfidentialityConstraintHandler.java:65)
	at io.undertow.security.handlers.AuthenticationMechanismsHandler.handleRequest(AuthenticationMechanismsHandler.java:60)
	at io.undertow.servlet.handlers.security.CachedAuthenticatedSessionHandler.handleRequest(CachedAuthenticatedSessionHandler.java:77)
	at io.undertow.security.handlers.NotificationReceiverHandler.handleRequest(NotificationReceiverHandler.java:50)
	at io.undertow.security.handlers.AbstractSecurityContextAssociationHandler.handleRequest(AbstractSecurityContextAssociationHandler.java:43)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.servlet.handlers.ServletInitialHandler.handleFirstRequest(ServletInitialHandler.java:247)
	at io.undertow.servlet.handlers.ServletInitialHandler.access$100(ServletInitialHandler.java:56)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:111)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:108)
	at io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:48)
	at io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43)
	at io.quarkus.undertow.runtime.UndertowDeploymentRecorder$9$1.call(UndertowDeploymentRecorder.java:587)
	at io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest(ServletInitialHandler.java:227)
	at io.undertow.servlet.handlers.ServletInitialHandler.handleRequest(ServletInitialHandler.java:152)
	at io.undertow.server.handlers.PathHandler.handleRequest(PathHandler.java:91)
	at io.quarkus.undertow.runtime.UndertowDeploymentRecorder$1.handleRequest(UndertowDeploymentRecorder.java:119)
	at io.undertow.server.Connectors.executeRootHandler(Connectors.java:290)
	at io.undertow.server.DefaultExchangeHandler.handle(DefaultExchangeHandler.java:18)
	at io.quarkus.undertow.runtime.UndertowDeploymentRecorder$5$1.run(UndertowDeploymentRecorder.java:413)
	at io.quarkus.runtime.CleanableExecutor$CleaningRunnable.run(CleanableExecutor.java:231)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2415)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1452)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at java.base/java.lang.Thread.run(Thread.java:836)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)


```

### Java version
```
openjdk 11.0.10 2021-01-19
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.10+9)
Eclipse OpenJ9 VM AdoptOpenJDK (build openj9-0.24.0, JRE 11 Linux amd64-64-Bit Compressed References 20210120_910 (JIT enabled, AOT enabled)
OpenJ9   - 345e1b09e
OMR      - 741e94ea8
JCL      - 0a86953833 based on jdk-11.0.10+9)
```

**Comment by mkouba** [comment]
> @mkouba this looks serious enough, could you have a look?

I will.

**Comment by mkouba** [comment]
@silviu-negoita what does the `com.connectis.myconnectis.adapters.web.infra.filters.AngularResourcesFilter` do? Does it have to be a normal scoped bean at all? It's not ideal, at least from the performance POV.

In the mean time, you can disable the monitoring functionality via `quarkus.arc.dev-mode.monitoring-enabled=false`.

**Comment by silviu-negoita** [comment]
@mkouba I will answer each question one at a time:

1. In this particular case, that filter is used to serve the FE (Angular app). I followed this official documentation from [here](https://quarkus.io/blog/quarkus-and-web-ui-development-mode/). 
```
@WebFilter(urlPatterns = "/*", asyncSupported = true)
@Priority(value = 300)
public class AngularResourcesFilter extends HttpFilter {
// code
}
```
But this is not the only source of exception. Other filters also throw this. 
2. It doesn't need. I removed @ApplicationScoped but this not fixed my issue
3. Indeed, using `quarkus.arc.dev-mode.monitoring-enabled=false` temporary fix my issue. 

 My hunch says that it is a filter chain issue.

**Comment by mkouba** [comment]
>     1. In this particular case, that filter is used to serve the FE (Angular app). I followed this official documentation...

I suspect the `asyncSupported = true` part...

> 2. It doesn't need. I removed @ApplicationScoped but this not fixed my issue

Yes, I've just verified that `@WebFilter` is registered as a custom bean defining annotation, i.e. all classes annotated with this annotation are registered as beans. `@ApplicationScoped` is indeed not the ideal scope for similar component though. `@Dependent` (default) or `@javax.inject.Singleton` would be a better match.

> 3. Indeed, using `quarkus.arc.dev-mode.monitoring-enabled=false` temporary fix my issue.

+1

**Comment by silviu-negoita** [comment]
@mkouba 

The `async` part is needed, or i get into:

```
2021-05-19 11:05:34,386 ERROR [io.und.req.io] (executor-thread-198) Exception handling request e3f77c19-4acf-4df4-85e6-fa41312527d0-2 to /myc/api/user/me: org.jboss.resteasy.spi.UnhandledException: java.lang.IllegalStateException: UT010026: Async is not supported for this request, as not all filters or Servlets were marked as supporting async
	at org.jboss.resteasy.core.ExceptionHandler.handleException(ExceptionHandler.java:381)
	at org.jboss.resteasy.core.SynchronousDispatcher.writeException(SynchronousDispatcher.java:218)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:519)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:261)
	at org.jboss.resteasy.core.SynchronousDispatcher$$Lambda$1412/0x00000000b8074250.run(Unknown Source)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:161)
	at org.jboss.resteasy.core.SynchronousDispatcher$$Lambda$1413/0x00000000b8076680.get(Unknown Source)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
	at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:164)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:247)
	at org.jboss.resteasy.plugins.server.servlet.ServletContainerDispatcher.service(ServletContainerDispatcher.java:249)
	at io.quarkus.resteasy.runtime.ResteasyFilter$ResteasyResponseWrapper.service(ResteasyFilter.java:70)
	at io.quarkus.resteasy.runtime.ResteasyFilter$ResteasyResponseWrapper.sendError(ResteasyFilter.java:76)
	at io.undertow.servlet.handlers.DefaultServlet.doGet(DefaultServlet.java:172)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:503)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:590)
	at io.undertow.servlet.handlers.ServletHandler.handleRequest(ServletHandler.java:74)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:129)
	at io.quarkus.resteasy.runtime.ResteasyFilter.doFilter(ResteasyFilter.java:31)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at com.connectis.myconnectis.adapters.web.infra.filters.AngularResourcesFilter.doFilter(AngularResourcesFilter.java:39)
	at com.connectis.myconnectis.adapters.web.infra.filters.AngularResourcesFilter_Subclass.doFilter$$superaccessor6(AngularResourcesFilter_Subclass.zig:1159)
	at com.connectis.myconnectis.adapters.web.infra.filters.AngularResourcesFilter_Subclass$$function$$20.apply(AngularResourcesFilter_Subclass$$function$$20.zig:47)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:54)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor.proceed(InvocationInterceptor.java:63)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor.monitor(InvocationInterceptor.java:49)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor_Bean.intercept(InvocationInterceptor_Bean.zig:521)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:41)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:41)
	at io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:32)
	at com.connectis.myconnectis.adapters.web.infra.filters.AngularResourcesFilter_Subclass.doFilter(AngularResourcesFilter_Subclass.zig:1094)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at io.undertow.servlet.handlers.FilterHandler.handleRequest(FilterHandler.java:84)
	at io.undertow.servlet.handlers.security.ServletSecurityRoleHandler.handleRequest(ServletSecurityRoleHandler.java:63)
	at io.undertow.servlet.handlers.ServletChain$1.handleRequest(ServletChain.java:68)
	at io.undertow.servlet.handlers.ServletDispatchingHandler.handleRequest(ServletDispatchingHandler.java:36)
	at io.undertow.servlet.handlers.RedirectDirHandler.handleRequest(RedirectDirHandler.java:67)
	at io.undertow.servlet.handlers.security.SSLInformationAssociationHandler.handleRequest(SSLInformationAssociationHandler.java:133)
	at io.undertow.servlet.handlers.security.ServletAuthenticationCallHandler.handleRequest(ServletAuthenticationCallHandler.java:57)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.security.handlers.AbstractConfidentialityHandler.handleRequest(AbstractConfidentialityHandler.java:46)
	at io.undertow.servlet.handlers.security.ServletConfidentialityConstraintHandler.handleRequest(ServletConfidentialityConstraintHandler.java:65)
	at io.undertow.security.handlers.AuthenticationMechanismsHandler.handleRequest(AuthenticationMechanismsHandler.java
....

```

Any alternatives?

**Comment by mkouba** [comment]
It should be fixed in the main branch.

**Comment by silviu-negoita** [comment]
@mkouba thanks for fast fix!

**Comment by mkouba** [comment]
> @mkouba thanks for fast fix!

It would be great if you could try to verify the fix ;-).

**Comment by silviu-negoita** [comment]
@mkouba I tested with current master branch(quarkus built locally) and my issue is fixed. This fix is included in latest `1.11.7.Final`?

**Comment by mkouba** [comment]
> I tested with current master branch(quarkus built locally) and my issue is fixed.

Great!

> This fix is included in latest 1.11.7.Final?

CC @gsmet
